### PR TITLE
Improve performance for physical plan creation with many columns

### LINKED ISSUE #12738: Execution plan creation performance degradation (is slower) compared to 0.40.0
I measured a performance of execution plan creation (and optimization)  using default physical planner.

0.40.0 vs current main (84ac4f9db09a88a9d24c33b26fe968dc5d8e0786).

To measure used the following test (put it into core/src/physical_planner.rs):
```rust
#[tokio::test]
async fn execution_plan_creation_bench() {
    let ctx = SessionContext::new();
    let mut fields = vec![];
    let mut columns = vec![];
    const COLUMNS_NUM: usize = 200;

    for i in 0..COLUMNS_NUM {
        fields.push(Field::new(format!("attribute{}", i), datafusion::arrow::datatypes::DataType::Int32, true));
        columns.push(Int32Array::from(vec![1]));
    }

    let batch = RecordBatch::try_new(
        SchemaRef::new(Schema::new(fields)),
        columns
            .into_iter()
            .map(|it| Arc::new(it) as Arc<dyn Array>)
            .collect(),
    )
    .unwrap();

    ctx.register_batch("t", batch).unwrap();

    let mut aggregates = String::new();
    for i in 0..COLUMNS_NUM {
        if i > 0 {
            aggregates.push_str(", ");
        }
        aggregates.push_str(format!("MAX(attribute{})", i).as_str());
    }

    /*
        SELECT max(attr0), ..., max(attrN) FROM t
    */
    let query = format!("SELECT {} FROM t", aggregates);
    let mut res = 0;

    let statement = ctx.state().sql_to_statement(&query, "Generic").unwrap();
    let plan = ctx.state().statement_to_plan(statement).await.unwrap();
    let planner = DefaultPhysicalPlanner::default();

    let mut duration_sum = 0.0;
    let mut durtaion_count = 0;

    for _ in 0..500 {
        let creation_start = std::time::Instant::now();
        let phys_plan = planner
            .create_physical_plan(&plan, &ctx.state())
            .await
            .unwrap();
        duration_sum += creation_start.elapsed().as_micros() as f64;
        durtaion_count += 1;
        res += phys_plan.children().len();
    }

    println!("mean = {}", duration_sum / (durtaion_count as f64));

    /* Not important. */
    println!("res [not important, just to not discard] = {}", res);
}
```

The query in the test is  `SELECT f(a0), ..., f(a_N) FROM t`, where f - some aggregate function.

The command to test:
```sh
cargo test --release --package datafusion --lib -- physical_planner::tests::execution_plan_creation_bench --exact --show-output
```

On the branch-40 I got the result:
```
mean = 1003.368
```

On the current main (84ac4f9) I got:
```
mean = 1286.524
```

But what is more interesting, when I increase the number of columns in the query (200 -> 1000), got:

branch-40:
```
mean = 8207.86
```

main (84ac4f9):
```
mean = 9836.384
```

So, the difference increases (it seems that the dependency is linear) with columns number growth. 

I create this issue to investigate: is such degradation inevitable? Or maybe there is a an obvious optimization that solves the problem.

### PR DESCRIPTION
**Pull request by askalt**
## Which issue does this PR close?

Closes #12738.

## Rationale for this change

I investigated the performance degradation in creating physical plans for queries with a large number of columns compared to version 40 and discovered the following:

1. The main time loss occurs during the cloning of plan nodes during optimizations. We can compare two flame graphs (for version 40 and version 42, attached) and see that in version 42, `enforce_distribution` spends additional time destroying the vector of AggregateFunExpr. It turns out that in patch f5c47fa274d53c1d524a1fb788d9a063bf5240ef, the Arc for storing aggregate expressions was removed. As a result, during calls to `with_new_children`, fairly heavy structures are being cloned and destroyed.

2. Some time was also spent on a new optimization: limit pushdown. This is optional and left up to the user to decide whether to use it, so there are no issues here.

This patch restores the Arc wrappers for storing aggregates in the physical plan nodes, and also adds a benchmark aimed at preventing the degradation from recurring.

Additionally:
As can be seen from the flame graphs, a significant amount of time is spent on creating ProjectionMapping. This mapping is only needed when there are eq properties, which are irrelevant for some plans, so it gets built unnecessarily, wasting time. I will raise a separate issue for this.

====

Flamegraphs here:

[flamegraphs.zip](https://github.com/user-attachments/files/17394457/flamegraphs.zip)

### CODE REVIEW

**Comment by askalt** [comment]
image

**Comment by alamb** [comment]
> As can be seen from the flame graphs, a significant amount of time is spent on creating ProjectionMapping. This mapping is only needed when there are eq properties, which are irrelevant for some plans, so it gets built unnecessarily, wasting time. I will raise a separate issue for this.

THank you

**Comment by alamb** [comment]
I have this on my list to review tomorrow if no one beats me to it

**Review by alamb — APPROVED** [review]
Thank you @askalt  -- I reviewed this PR carefully and Arc'ing the structure makes quite a bit of sense. 

Here is a proposal to reduce copying even more: https://github.com/askalt/datafusion/pull/1

**Comment by alamb** [comment]
> Additionally:
As can be seen from the flame graphs, a significant amount of time is spent on creating ProjectionMapping. This mapping is only needed when there are eq properties, which are irrelevant for some plans, so it gets built unnecessarily, wasting time. I will raise a separate issue for this.

It would be great to file an issue for this -- I am happy to do so if you like

**Comment by askalt** [comment]
> > Additionally:
> > As can be seen from the flame graphs, a significant amount of time is spent on creating ProjectionMapping. This mapping is only needed when there are eq properties, which are irrelevant for some plans, so it gets built unnecessarily, wasting time. I will raise a separate issue for this.
> 
> It would be great to file an issue for this -- I am happy to do so if you like
Yes, I like, thank you!

File: datafusion/core/benches/physical_plan.rs [hunk 1]
```diff
@@ -183,5 +187,67 @@ fn criterion_benchmark(c: &mut Criterion) {
     }
 }
 
-criterion_group!(benches, criterion_benchmark);
+/// Aimed at tracking inefficiencies at the stage of creating/optimizing a physical plan.
+fn bench_creation_many_columns(c: &mut Criterion) {
```
[the remark this answers is not in the document]
Comment (askalt) [reply datafusion/core/benches/physical_plan.rs]: Done.

**Comment by alamb** [comment]
Thanks again @askalt  -- I am now working on the follow on ticket

**Comment by alamb** [comment]
Filed https://github.com/apache/datafusion/issues/13015 and a follow on issue here https://github.com/apache/datafusion/pull/13012

**Pull request merged**